### DISCUSSION TITLE
Do I have to add `crossterm` crate manually?

### DISCUSSION THREAD

**Discussion by liuhq** [ratatui/ratatui#1585] [Q&A] [answered]
In the doc site, the tutorials guides me initializing a ratatui project which should add `ratatui` and `crossterm` crates. But I found that the project can be built and run without adding `crossterm` to the dependencies manually.

(and I use the `crossterm`'s fields, functions and structures through `ratatui::{ crossterm::event::{ self, ... }}`)

Is there something problems that I haven't noticed? or is it okay too?

**Comment by joshka** [comment, marked as answer]
TLDR - both work. I prefer the explicit import of crossterm rather than importing from the re-export.

We added a re-export of the crossterm library at `ratatui::crossterm` recently to make it easier to ensure to ensure that the correct version is used. This is a pretty common practice for libs. That said, I think it was probably a (small) mistake to import crossterm from all the example imports. That is to say, you can just use `ratatui::crossterm::event` if you'd like, but using `crossterm::event` seems more idiomatic for an app.

The one place where this becomes a problem is when you have multiple semver incompatible versions available in your dependency tree. Version 0.x and lib 0.x+1 are treated as two entirely different dependencies despite having the same names. We're working to help avoid this problem more generally in https://github.com/ratatui/ratatui/issues/1388.

**Reply by orhun** [reply]
We should probably update the examples/docs to remove the explicit import of crossterm wherever it is redundant and make a note about this in the documentation. What do you think?

I'm wondering if @liuhq would be up for contributing wrt this :)

**Reply by joshka** [reply]
More the other way around IMO. Include it explicitly seems to make sense.
For the tutorials I'd like to get these all using the templates, so adding it to your cargo.toml is irrelevant (it's already there).

**Reply by liuhq** [reply]
@joshka Okay, thx! I think I get it. Currently, using `ratatui::crossterm::event` works fine. Whether to explicitly use `crossterm::event` depends more on personal preference for the proj's structure.

@orhun I think that we need to maintain a consistent style for beginners here, regardless of whether `crossterm::event` is used explicitly. Otherwise, the different styles ([re-export](https://ratatui.rs/tutorials/json-editor/main/#application-pre-run-steps) and [explicit](https://ratatui.rs/tutorials/counter-app/basic-app/#main-imports)) in the documentation might cause confusion. Then, what the style should we choose? Explicitly `crossterm::event`?

**Reply by joshka** [reply]
I prefer being explicit imports in our docs, examples etc. Module re-exports are something that not a lot of other languages tend to have, so this seems like an easy place to trip newbies up on for no real gain. Which you choose makes no real difference to the compiler, only to the clarity of explanation.
